### ISSUE TITLE
Can't get transaction hash before transaction is deployed

### ISSUE THREAD

**Issue by marekkirejczyk** [trufflesuite/truffle-contract#42]
I can't get transaction hash before a transaction is mined (or contract deployed)
One example of this being useful is to show a link to a blockchain explorer, as the transaction is being mined.

The current state of API:
```
MyContract.new(arg1, arg2, {txArg1, txArg2}).then((t) => {
  console.log(“Transction mined!”) 
});
```

**Proposed improvement:**
Variant 1:
```
MyContract.new(arg1, arg2, {txArg1, txArg2}, (hash) => { 
   console.log(`Waitng for ${hash} transaction to be mined`)
}).then((t) => {
    console.log(“Contract deployed!”);
}
```

Variant 2:
```
MyContract.fireCallbackOnTransactionSend = true;
MyContract.new(arg1, arg2, {txArg1, txArg2}).then((t) => {
  if (t.address)
    console.log(“Contract deployed!”);
  else
    console.log(“Waiting to be mined: ”: t.transactionHash);
}
```

Variant 3:
```
MyContract.onTransactionHash = (transaction) => console.log(“Waiting to be mined: ”: transaction.hash);
MyContract.new(arg1, arg2, {txArg1, txArg2}).then((contract) => {
    console.log(“Contract deployed!”);
}
```

Feedback welcome. I'll be happy to do a PR :)

**Comment by ajsantander** [comment]
This would be super handy yes.

**Comment by marekkirejczyk** [comment]
As I was thinking about it a nicer interface would be:
```
MyContract.new(arg1, arg2, {txArg1, txArg2}).onHash((hash) => {
  console.log(“Transaction hash:”, hash) 
}).onMined((t) => {
  console.log(“Transaction mined!”) 
});
```

**Comment by garyng2000** [comment]
indeed. I am looking for the some functionality, thinking about not using truffle-contract but going back to the good old callback of web3.
my use case is to watch the event for that particular transaction for UI purpose.

**Comment by marekkirejczyk** [comment]
Can I get confirmation from maintainer it will be merged i

**Comment by marekkirejczyk** [comment]
Closed by mistake.

**Comment by buybackoff** [comment]
Hash is very needed indeed before promises return - not only for contract deployment but for any transaction. TestRPC is fast, but on e.g. Ropsten waiting for promises is often not reliable (not due to this library, but e.g. Jest timeout or page reload), need ACK that transaction is sent as soon as it is out, so that I could save Tx hash and long-poll later even if couldn't wait for a promise to return.

**Comment by cag** [comment]
Yikes, looks like there are two PRs designed for addressing this issue. #62 uses a callback on the returned thenable. #73 factors the synchronization logic out into its own thing.

**Comment by cgewecke** [comment]
This feature is included in `truffle-contract`'s  upgrade to `web3 1.0`, now available as an experimental nightly build here: [darq-truffle@next](https://www.npmjs.com/package/darq-truffle). 
```shell
$ npm install -g darq-truffle@next
$ darq-truffle test # Example command
```

Usage docs [here](https://github.com/trufflesuite/truffle-contract/blob/web3-one-readme/README.md)
